File: examples/Csharp.ExampleApplication/Hooks/MigrationActionCompleted/LogMigrationActionsHook.cs
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Tableau.Migration.Engine.Actions;
using Tableau.Migration.Engine.Hooks;

namespace Csharp.ExampleApplication.Hooks.MigrationActionCompleted
{
    #region class
    public class LogMigrationActionsHook : IMigrationActionCompletedHook
    {
        private readonly ILogger<LogMigrationActionsHook> _logger;

        public LogMigrationActionsHook(ILogger<LogMigrationActionsHook> logger)
        {
            _logger = logger;
        }

        public Task<IMigrationActionResult?> ExecuteAsync(IMigrationActionResult ctx, CancellationToken cancel)
        {
            if (ctx.Success)
            {
                _logger.LogInformation("Migration action completed successfully.");
            }
            else
            {
                _logger.LogWarning(
                    "Migration action completed with errors:{NewLine}{Errors}",
                    Environment.NewLine,
                    String.Join(Environment.NewLine, ctx.Errors.Select(e => e.ToString())));
            }

            return Task.FromResult<IMigrationActionResult?>(ctx);
        }
    }
    #endregion
}

File: examples/Python.ExampleApplication/Hooks/migration_action_completed/log_migration_actions_hook.py
import logging
from tableau_migration import(
    MigrationActionCompletedHookBase,
    IMigrationActionResult
    )


class LogMigrationActionsHook(MigrationActionCompletedHookBase):
    def __init__(self) -> None:
        super().__init__()
        
        # Create a logger for this class
        self._logger = logging.getLogger(__name__)
        
    def execute(self, ctx: IMigrationActionResult) -> IMigrationActionResult:
        if(ctx.success):
            self._logger.info("Migration action completed successfully.")
        else:
            all_errors = "\n".join(ctx.errors)
            self._logger.warning("Migration action completed with errors:\n%s", all_errors)
            
        return None
